gpt4all-chat/mistral.cpp

```diff
@@ -0,0 +1,259 @@
+#include "mistral.h"
+
+#include <string>
+#include <vector>
+#include <iostream>
+
+#include <QCoreApplication>
+#include <QThread>
+#include <QEventLoop>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonArray>
+
+//#define DEBUG
+
+Mistral::Mistral()
+    : QObject(nullptr)
+    , m_modelName("tiny")
+    , m_responseCallback(nullptr)
+{
+}
+
+size_t Mistral::requiredMem(const std::string &modelPath, int n_ctx, int ngl)
+{
+    Q_UNUSED(modelPath);
+    Q_UNUSED(n_ctx);
+    Q_UNUSED(ngl);
+    return 0;
+}
+
+bool Mistral::loadModel(const std::string &modelPath, int n_ctx, int ngl)
+{
+    Q_UNUSED(modelPath);
+    Q_UNUSED(n_ctx);
+    Q_UNUSED(ngl);
+    return true;
+}
+
+void Mistral::setThreadCount(int32_t n_threads)
+{
+    Q_UNUSED(n_threads);
+    qt_noop();
+}
+
+int32_t Mistral::threadCount() const
+{
+    return 1;
+}
+
+Mistral::~Mistral()
+{
+}
+
+bool Mistral::isModelLoaded() const
+{
+    return true;
+}
+
+// All three of the state virtual functions are handled custom inside of chatllm save/restore
+size_t Mistral::stateSize() const
+{
+    return 0;
+}
+
+size_t Mistral::saveState(uint8_t *dest) const
+{
+    Q_UNUSED(dest);
+    return 0;
+}
+
+size_t Mistral::restoreState(const uint8_t *src)
+{
+    Q_UNUSED(src);
+    return 0;
+}
+
+void Mistral::prompt(const std::string &prompt,
+                     const std::string &promptTemplate,
+                     std::function<bool(int32_t)> promptCallback,
+                     std::function<bool(int32_t, const std::string&)> responseCallback,
+                     std::function<bool(bool)> recalculateCallback,
+                     PromptContext &promptCtx,
+                     bool special,
+                     std::string *fakeReply) {
+
+    Q_UNUSED(promptCallback);
+    Q_UNUSED(recalculateCallback);
+    Q_UNUSED(special);
+    Q_UNUSED(fakeReply);
```

Comment:
This PR inherits bugs that have been fixed in #2051. Ideally, you would find a way to share more code between chatgpt.cpp and mistral.cpp, so changes to them don't need to be copy-pasted back and forth.

Reply:
I do plan on merging both of them and implementing a better way to handle apis. Is it out of the scope of the issue or should I implement it in this PR ?